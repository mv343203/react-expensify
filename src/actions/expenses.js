import {v4 as uuid} from 'uuid';


//this creates an action to interact with store with default properties 
//names the type of action
//this actions is stating to add the expense to the store 
//use the default properties listed first if nothing is provided
//we will be adding the object of expense on the array 
export const addExpense = (
    {
        description = '', 
        note = '', 
        amount = 0,
        createdAt = 0
      } = {}
    ) => ({
         type: 'ADD_EXPENSE',
         expense: {
           id: uuid(),
           description,
           note,
           amount,
           createdAt
    }
});

//creates action to interact with store by fetching id property
//fetches the id property  
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});
