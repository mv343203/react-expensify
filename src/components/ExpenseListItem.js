import React from 'react';
import { Link } from 'react-router-dom'


//destructure props object to access to individual data  from the props object 
//we will then import into the ExpenseList 
//access the store by adding dispatch into the spread and id to grab specefic id 
const ExpenseListItem = ({id, description, amount, createdAt}) => (
  <div>
    <Link to = {`/edit/${id}`}>  
      <h1>{description}</h1>
    </Link>
    
    <p>{amount} - {createdAt} </p>
    
  </div>
);

//default 
export default ExpenseListItem;