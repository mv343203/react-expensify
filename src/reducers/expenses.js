//sets a default state for the store to return 
const expensesReducerDefaultState = [];

//reducer which describes how the state changes in response to action
//call actions types with cases 
export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            //returns new array with subset values
            //if function returns true item will be kept
            //if false item will be removed
            //we destructured the expense array and just got its id
            //if id does not = action id then id will be kept
            //if = then id is removed by filter
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_EXPENSE':
            //iterates over array in expenses
            return state.map((expense) => {
                //checks to match ids and if no changes made
                if (expense.id === action.id) {
                    //if there is a match spread operator and update expense with updates
                    return {
                        ...expense,
                        ...action.updates
                    };

                } else {
                    return expense;
                };
            }); 
        default:
            return state;

    }
};
