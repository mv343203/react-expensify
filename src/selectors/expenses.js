//Get visible expenses
//expenses is the array it will sorting 
//filtering is what willbe applied - destructured the filtering objectit to get properties
//using timestamps which start at 1970 January 1
import moment from 'moment';

export default(expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
      if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === 'amount') {
        return a.amount < b.amount ? 1 : -1;
      }
    });
  };