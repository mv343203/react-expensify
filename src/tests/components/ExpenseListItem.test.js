import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import toJSON from 'enzyme-to-json';
import expenses from '../fixtures/expenses';


test('should render ExpenseListItem with expenses', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
});