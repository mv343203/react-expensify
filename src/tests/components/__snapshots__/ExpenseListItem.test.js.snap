// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with expenses 1`] = `
<div>
  <[object Object]
    to="/edit/1"
  >
    <h1>
      Gum
    </h1>
  </[object Object]>
  <p>
    195
     - 
    0
     
  </p>
</div>
`;
